crates/core/src/factories/core_publish.rs

```diff
@@ -0,0 +1,247 @@
+    ) -> Self {
+        Self::spawn_tasks(config, space_id, fetch, transport)
+    }
+}
+
+impl Publish for CorePublish {
+    fn publish_ops(
+        &self,
+        op_ids: Vec<OpId>,
+        target: Url,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Insert requests into publish queue.
+            for op_id in op_ids {
+                if let Err(err) = self
+                    .outgoing_publish_ops_tx
+                    .send((op_id, target.clone()))
+                    .await
+                {
+                    tracing::warn!(
+                        "could not insert ops into ops publish queue: {err}"
+                    );
+                }
+            }
+
+            Ok(())
+        })
+    }
+}
+
+impl CorePublish {
+    pub fn spawn_tasks(
+        _config: CorePublishConfig,
+        space_id: SpaceId,
+        fetch: DynFetch,
+        transport: DynTransport,
+    ) -> Self {
+        // Create a queue to process outgoing op publishes. Publishes are sent to peers.
+        let (outgoing_publish_ops_tx, outgoing_publish_ops_rx) =
+            channel::<OutgoingPublishOps>(16_384);
+
+        // Create a queue to process incoming op publishes. Incoming publishes are added to the
+        // fetch queue
+        let (incoming_publish_ops_tx, incoming_publish_ops_rx) =
+            channel::<IncomingPublishOps>(16_384);
+
+        let mut tasks = Vec::new();
+
+        // Spawn outgoing publish ops task.
+        let outgoing_publish_ops_task =
+            tokio::task::spawn(CorePublish::outgoing_publish_ops_task(
+                outgoing_publish_ops_rx,
+                space_id.clone(),
+                transport.clone(),
+            ))
+            .abort_handle();
+        tasks.push(outgoing_publish_ops_task);
+
+        // Spawn incoming publish ops task.
+        let incoming_publish_ops_task =
+            tokio::task::spawn(CorePublish::incoming_publish_ops_task(
+                incoming_publish_ops_rx,
+                fetch,
+            ))
+            .abort_handle();
+        tasks.push(incoming_publish_ops_task);
+
+        // Register transport module handler for incoming op and agent publishes.
+        let message_handler = Arc::new(PublishMessageHandler {
+            incoming_publish_ops_tx,
+        });
+
+        transport.register_module_handler(
+            space_id.clone(),
+            PUBLISH_MOD_NAME.to_string(),
+            message_handler.clone(),
+        );
+
+        Self {
+            outgoing_publish_ops_tx,
+            tasks,
+        }
+    }
+
+    async fn outgoing_publish_ops_task(
+        mut outgoing_publish_ops_rx: Receiver<OutgoingPublishOps>,
+        space_id: SpaceId,
+        transport: DynTransport,
+    ) {
+        while let Some((op_id, peer_url)) = outgoing_publish_ops_rx.recv().await
+        {
+            // Send ops publish message to peer.
+            let data = serialize_publish_ops_message(vec![op_id.clone()]);
```

Comment:
This is where it looks slightly odd to me to publish one op at a time. The protocol was set up to publish a list but we're always publishing one thing